src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -391,37 +434,66 @@ ucs_status_t uct_tcp_sockcm_cm_ep_conn_notify(uct_ep_h ep)
     return status;
 }
 
-static ucs_status_t uct_tcp_sockcm_ep_pack_priv_data(uct_tcp_sockcm_ep_t *cep)
+static ucs_status_t
+uct_tcp_sockcm_ep_invoke_resolve_cb(uct_tcp_sockcm_ep_t *cep,
+                                    const char *ifname)
 {
-    char ifname_str[UCT_DEVICE_NAME_MAX];
-    uct_tcp_sockcm_priv_data_hdr_t *hdr;
-    size_t priv_data_ret;
+    uct_cm_ep_resolve_args_t resolve_args;
     ucs_status_t status;
-    uct_cm_ep_priv_data_pack_args_t pack_args;
 
-    /* get interface name associated with the connected client fd */
-    status = ucs_sockaddr_get_ifname(cep->fd, ifname_str, sizeof(ifname_str));
-    if (UCS_OK != status) {
-        goto out;
-    }
+    resolve_args.field_mask = UCT_CM_EP_RESOLVE_ARGS_FIELD_DEV_NAME;
+    ucs_strncpy_safe(resolve_args.dev_name, ifname, UCT_DEVICE_NAME_MAX);
+    status      = uct_cm_ep_resolve_cb(&cep->super, &resolve_args);
+    cep->state |= (status == UCS_OK) ? UCT_TCP_SOCKCM_EP_RESOLVED :
+                  UCT_TCP_SOCKCM_EP_RESOLVE_CB_FAILED;
+    return status;
+}
+
+static ucs_status_t
+uct_tcp_sockcm_ep_invoke_pack_cb(uct_tcp_sockcm_ep_t *cep,
+                                 const char *ifname)
+{
+    uct_cm_ep_priv_data_pack_args_t pack_args;
+    uct_tcp_sockcm_priv_data_hdr_t *hdr;
+    ucs_status_t status;
 
-    hdr                  = (uct_tcp_sockcm_priv_data_hdr_t*)cep->comm_ctx.buf;
     pack_args.field_mask = UCT_CM_EP_PRIV_DATA_PACK_ARGS_FIELD_DEVICE_NAME;
-    ucs_strncpy_safe(pack_args.dev_name, ifname_str, UCT_DEVICE_NAME_MAX);
+    ucs_strncpy_safe(pack_args.dev_name, ifname, UCT_DEVICE_NAME_MAX);
 
+    ucs_assert(cep->comm_ctx.offset == 0);
+    hdr = (uct_tcp_sockcm_priv_data_hdr_t*)cep->comm_ctx.buf;
     status = uct_cm_ep_pack_cb(&cep->super, cep->super.user_data, &pack_args,
                                hdr + 1,
                                uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len,
-                               &priv_data_ret);
+                               &hdr->length);
     if (status != UCS_OK) {
         cep->state |= UCT_TCP_SOCKCM_EP_PACK_CB_FAILED;
-        goto out;
+        return status;
     }
 
-    hdr->length          = priv_data_ret;
     hdr->status          = (uint8_t)UCS_OK;
     cep->comm_ctx.length = sizeof(*hdr) + hdr->length;
     cep->state          |= UCT_TCP_SOCKCM_EP_PRIV_DATA_PACKED;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_resolve(uct_tcp_sockcm_ep_t *cep)
```

Comment:
if this function does either resolve or pack then its name should reflect it.
uct_tcp_sockcm_ep_resolve_or_pack ?

Reply:
this function handles lower level "resolve" event, maybe rename it to `uct_tcp_sockcm_ep_resolve_handler`? also onetime we want to deprecate and delete old API, so I would like to reflect naming to "preferable" path

Reply:
if this function does pack, why is it 'lower level "resolve" event' ?